### PR TITLE
fix: Include installed modules in all database definitions.

### PR DESCRIPTION
**Pull request by Isakdl**
# Changes

Make installed modules in database definition none nullable.
Always load installed modules into the database definition.
Recreates all migrations for all projects.

Related: #1499

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

_If you have done any breaking changes, make sure to outline them here, so that they can be included in the notes for the next release._

### CODE REVIEW

File: tools/serverpod_cli/lib/src/create/database_setup.dart [hunk 1]
```diff
@@ -6,9 +6,18 @@ import 'package:serverpod_cli/src/logger/logger.dart';
 import 'package:serverpod_cli/src/util/command_line_tools.dart';
 
 class DatabaseSetup {
-  static Future<bool> createDefaultMigration(Directory dir, String name) async {
+  static Future<bool> createDefaultMigration(
+    Directory dir,
+    String name,
+  ) async {
     log.debug('Creating default migration.');
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/create/database_setup.dart:13]: Maybe we should rephrase this as `Creating initial migration`? since we will actually create one but we won't have the tag.

File: examples/auth_example/auth_example_server/generated/migration/migrations/serverpod/20231124134350/definition.json [hunk 2]
```diff
@@ -1367,7 +1367,7 @@
       "managed": true
     }
   ],
-  "priority": 0,
-  "installedModules": null,
+  "priority": null,
```
Comment (SandPod) [inline examples/auth_example/auth_example_server/generated/migration/migrations/serverpod/20231124134350/definition.json:1370]: Question: Shouldn't priority still be 0? Until we remove it in PR 3?

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 3]
```diff
@@ -66,6 +69,7 @@ class MigrationGenerator {
     );
     try {
       return await MigrationVersion.load(
+        moduleName: module,
         versionName: versionName,
         migrationsDirectory: migrationsDirectory,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:74]: Suggestion(Non-blocking): Move the migrationsDirectory construction into the load method.

Since we now pass in the module name as well we might as well do the construction for the migrations directory inside of `MigrationVersion.load()`.

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 4]
```diff
@@ -122,19 +140,34 @@ class MigrationGenerator {
...
     bool write = true,
+    required GeneratorConfig config,
   }) async {
     var migrationRegistry = MigrationRegistry.load(
       migrationsProjectDirectory,
     );
 
+    var protocols =
+        await ProtocolHelper.loadProjectYamlProtocolsFromDisk(config);
+    var entityDefinitions = StatefulAnalyzer(protocols).validateAll();
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:151]: Suggestion: Print warnings and errors and stop if any are found.

This should be found in a e2e test as well. But it looks like the printing and stopping has been removed this was moved from old class.

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 5]
```diff
@@ -101,14 +105,28 @@ class MigrationGenerator {
...
   Future<DatabaseDefinition> _getSourceDatabaseDefinition(
     String? latestVersion,
     int priority,
+    List<DatabaseMigrationVersion> installedModules,
   ) async {
     if (latestVersion == null) {
       return DatabaseDefinition(
         tables: [],
         priority: priority,
+        installedModules: installedModules,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:129]: Suggestion: Make this an empty list.

If we don't find a source database definition we should assume there is no previous migration. And in that case nothing should have been "installed". This should therefore be empty.

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 6]
```diff
@@ -122,19 +140,34 @@ class MigrationGenerator {
...
       priority,
+      installedModules,
     );
 
-    var dstDatabase = await generateDatabaseDefinition(
-      directory: directory,
-      priority: priority,
+    var dstDatabase = createDatabaseDefinitionFromEntities(
+      entityDefinitions,
+      config,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:170]: Suggestion: Still pass in priority here.

This PR shouldn't change the use of priority. That work should be done in a separate PR when the new SQL system is in place.

File: tools/serverpod_cli/lib/src/database/create_definition.dart [hunk 7]
```diff
@@ -1,15 +1,19 @@
+import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/analyzer/entities/definitions.dart';
 import 'package:serverpod_shared/serverpod_shared.dart';
 import 'package:serverpod_service_client/serverpod_service_client.dart';
 
 /// Create the target [DatabaseDefinition] based on the [serializableEntities].
 DatabaseDefinition createDatabaseDefinitionFromEntities(
-    List<SerializableEntityDefinition> serializableEntities) {
+  List<SerializableEntityDefinition> serializableEntities,
+  GeneratorConfig config,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/database/create_definition.dart:9]: Suggestion: Pass in module name and installed versions as separate parameters.

Since GeneratorConfig contains so much information for the project it becomes hard to detect what is actually required from it in this place. I much destructing the object as early as possible and passing down what is required to the different places. That creates less places to inspect and understand if we would like to change it in the future.

File: tools/serverpod_cli/lib/src/create/database_setup.dart [hunk 8]
```diff
@@ -6,9 +6,18 @@ import 'package:serverpod_cli/src/logger/logger.dart';
...
 
+    var config = await GeneratorConfig.load(dir.path);
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/create/database_setup.dart:15]: Question: What config is loaded here? The one in the newly created project?

File: tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart [hunk 9]
```diff
@@ -27,6 +40,7 @@ class MigrationVersionBuilder {
     return this;
   }
 
+  /// Should include your self in the installedModules.
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart:43]: Suggestion: Remove comment.

This seems like an instruction for when you were writing the code. If it should stay could you rephrase it so that it is understandable for future devs.

Comment (SandPod) [reply tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart:43]: Maybe:
```suggestion
  /// The [DatabaseDefinition] should always include its own module in its installedModules.
```

**Review by SandPod — CHANGES_REQUESTED** [review]
Nice to see the addition in the model.

First round of comments.
We should separate out the removal of Priority completely its own PR.

Aslo detected some severe regressions.

**Review by SandPod — CHANGES_REQUESTED** [review]
Nice to see the addition in the model.

First round of comments.
We should separate out the removal of Priority completely its own PR.

Aslo detected some severe regressions.

File: tools/serverpod_cli/lib/src/create/database_setup.dart (hunk 8 above)
Comment (Isakdl) [reply tools/serverpod_cli/lib/src/create/database_setup.dart:15]: yes

File: tools/serverpod_cli/lib/src/migrations/generator.dart (hunk 3 above)
Comment (Isakdl) [reply tools/serverpod_cli/lib/src/migrations/generator.dart:74]: Not sure I follow here.

File: tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart (hunk 9 above)
Comment (Isakdl) [reply tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart:43]: But that seems like you should only include other modules?! The point here is you have to add your own project in the modules as well.

File: examples/chat/chat_server/generated/migration/migrations/chat/20231124134459/definition.json [hunk 10]
```diff
@@ -50,7 +50,26 @@
       "managed": true
     }
   ],
-  "priority": 2,
-  "installedModules": null,
+  "priority": null,
```
Comment (SandPod) [inline examples/chat/chat_server/generated/migration/migrations/chat/20231124134459/definition.json:53]: chore: generate all projects.

Priority is still null.

File: packages/serverpod/lib/src/generated/protocol.dart [hunk 11]
```diff
@@ -1806,13 +1812,6 @@ class Protocol extends _i1.SerializationManagerServer {
           .map((e) => deserialize<_i59.TableDefinition>(e))
           .toList() as dynamic;
     }
-    if (t == _i1.getType<List<_i59.DatabaseMigrationVersion>?>()) {
```
Comment (SandPod) [inline packages/serverpod/lib/src/generated/protocol.dart:1809]: Question: Hmm, why is this removed?

File: tools/serverpod_cli/lib/src/migrations/generator.dart (hunk 3 above)
Comment (SandPod) [reply tools/serverpod_cli/lib/src/migrations/generator.dart:74]: On the lines above we create the migrationsDirectory. But we have all the information to do this internally now inside of the load method. We could just pass in the migrationBaseDirectory and then MigrationVersion could construct the path to the designated directory.

(The first thing we do inside of `load(...)` is actually to build another directory path using the version name. So the building on the lines above here is almost completetly redundant.)

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 12]
```diff
@@ -101,6 +105,18 @@ class MigrationGenerator {
...
 
+  Future<List<MigrationVersion>> getAllMigrationVersions(
+    List<String> modules,
+  ) async {
+    var versions = <Future<MigrationVersion?>>[];
+    for (var module in modules) {
+      var moduleVersions = getLatestMigrationVersion(module);
+      versions.add(moduleVersions);
+    }
+    var resolved = await Future.value(versions);
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:116]: NEAT! 🚀

File: tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart [hunk 13]
```diff
@@ -2,11 +2,13 @@ import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/database/create_definition.dart';
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
 import 'package:serverpod_cli/src/test_util/builders/foreign_relation_definition_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
 import 'package:serverpod_cli/src/test_util/builders/serializable_entity_field_definition_builder.dart';
 import 'package:serverpod_service_client/serverpod_service_client.dart';
 import 'package:test/test.dart';
 
 void main() {
+  var config = GeneratorConfigBuilder().build();
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart:11]: Suggestion: Remove the GeneratorConfigBuilder and just define the name and modules directly in the test.

Since we don't really need the object anymore it is more explicit if we create the things we need for the test.

File: tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart [hunk 14]
```diff
@@ -2,9 +2,12 @@ import 'package:recase/recase.dart';
 import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/database/create_definition.dart';
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
 import 'package:test/test.dart';
 
 void main() {
+  var config = GeneratorConfigBuilder().build();
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart:9]: Suggestion: Remove the generator config builder and define what we need directly.

**Review by SandPod — COMMENTED** [review]
Couple of comments still, I think there was a push mid review so some might be outdated.

File: tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart (hunk 9 above)
Comment (SandPod) [reply tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart:43]: Yeah, might not be the right phrasing either. But something that makes it a bit more clear what should be included.

/// The [DatabaseDefinition] should always include a reference to itself as a module in its installedModules.

File: packages/serverpod/lib/src/generated/protocol.dart (hunk 11 above)
Comment (Isakdl) [reply packages/serverpod/lib/src/generated/protocol.dart:1809]: probably because it is none-nullable now.

File: tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart [hunk 15]
```diff
@@ -2,6 +2,7 @@ import 'package:recase/recase.dart';
 import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/database/create_definition.dart';
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
+
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart:5]: NIT: Remove empty line.

File: tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart [hunk 16]
```diff
@@ -2,6 +2,7 @@ import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/database/create_definition.dart';
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
 import 'package:serverpod_cli/src/test_util/builders/foreign_relation_definition_builder.dart';
+
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart:5]: NIT: remove empty line.

Comment (SandPod) [reply tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart]: NIT:
```suggestion
```

Comment (SandPod) [reply tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart]: NIT
```suggestion
```

File: tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart [hunk 17]
```diff
@@ -4,18 +4,31 @@ import 'package:serverpod_cli/src/migrations/generator.dart';
...
-  );
-  Directory _migrationsDirectory = Directory.current;
+  late DatabaseDefinition _databaseDefinition;
+  Directory _versionDirectory = Directory.current;
+
+  MigrationVersionBuilder() {
+    _databaseDefinition = DatabaseDefinition(
+      installedModules: [
+        DatabaseMigrationVersion(module: _moduleName, version: _versionName)
+      ],
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/test_util/builders/migration_version_builder.dart]: Suggestion: Add serverpod.

Serverpod is included in most scenarios.

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 18]
```diff
@@ -58,17 +61,11 @@ class MigrationGenerator {
...
-      ),
-    );
     try {
       return await MigrationVersion.load(
-        versionName: versionName,
-        migrationsDirectory: migrationsDirectory,
-      );
+          moduleName: module,
+          versionName: versionName,
+          baseDir: migrationsBaseDirectory);
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:68]: NIT:
```suggestion
          baseDir: migrationsBaseDirectory,
        );
```

File: tools/serverpod_cli/lib/src/migrations/generator.dart [hunk 19]
```diff
@@ -364,24 +396,28 @@ class MigrationGenerator {
...
   final String versionName;
   final DatabaseMigration migration;
   final DatabaseDefinition databaseDefinition;
   int get priority => migration.priority;
 
   static Future<MigrationVersion> load({
+    required String moduleName,
     required String versionName,
-    required Directory migrationsDirectory,
+    required Directory baseDir,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/migrations/generator.dart:417]: Suggestion(NIT): 
```suggestion
    required Directory migrationDirectory,
```

Since this is the directory that holds all the migrations.

**Review by SandPod — APPROVED** [review]
LGTM! Couple of last NITs that would be good to fix.

**Review by SandPod — COMMENTED** [review]
LGTM! Couple of last NITs that would be good to fix.
